src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/ByValueMarshalKindSupportDescriptor.cs

```diff
@@ -0,0 +1,137 @@
+    /// </summary>
+    public record ByValueMarshalKindSupportDescriptor(
+        ByValueMarshalKindSupport InSupport, string? InSupportDetails,
+        ByValueMarshalKindSupport OutSupport, string? OutSupportDetails,
+        ByValueMarshalKindSupport InOutSupport, string? InOutSupportDetails)
+    {
+        /// <summary>
+        /// A default <see cref="ByValueMarshalKindSupportDescriptor"/> for by value parameters. [In] is allowed, but unnecessary. Out is not allowed.
+        /// </summary>
+        public static readonly ByValueMarshalKindSupportDescriptor ValueTypeParameterDefault = new ByValueMarshalKindSupportDescriptor(
+            InSupport: ByValueMarshalKindSupport.Unnecessary, InSupportDetails: SR.InAttributeOnlyIsDefault,
+            OutSupport: ByValueMarshalKindSupport.NotSupported, OutSupportDetails: SR.OutAttributeNotSupportedOnByValueValueTypeParameters,
+            InOutSupport: ByValueMarshalKindSupport.NotSupported, InOutSupportDetails: SR.OutAttributeNotSupportedOnByValueValueTypeParameters);
+
+        /// <summary>
+        /// A default <see cref="ByValueMarshalKindSupportDescriptor"/> for by value parameters. [In] is allowed, but unnecessary. Out is not allowed.
+        /// </summary>
+        public static readonly ByValueMarshalKindSupportDescriptor ReferenceTypeParameterDefault = new ByValueMarshalKindSupportDescriptor(
+            InSupport: ByValueMarshalKindSupport.Unnecessary, InSupportDetails: SR.InAttributeOnlyIsDefault,
+            OutSupport: ByValueMarshalKindSupport.NotSupported, OutSupportDetails: SR.OutAttributeNotSupportedOnByValueReferenceTypeParameters,
+            InOutSupport: ByValueMarshalKindSupport.NotSupported, InOutSupportDetails: SR.OutAttributeNotSupportedOnByValueReferenceTypeParameters);
+
+        /// <summary>
+        /// A default <see cref="ByValueMarshalKindSupportDescriptor"/> for by reference parameters. [In] is allowed, but unnecessary. Out is allowed.
+        /// </summary>
+        public static readonly ByValueMarshalKindSupportDescriptor ArrayParameterDefault = new ByValueMarshalKindSupportDescriptor(
+            InSupport: ByValueMarshalKindSupport.Unnecessary, InSupportDetails: SR.InAttributeOnlyIsDefault,
+            OutSupport: ByValueMarshalKindSupport.Supported, OutSupportDetails: null,
+            InOutSupport: ByValueMarshalKindSupport.Supported, InOutSupportDetails: null);
+
+        /// <summary>
+        /// A default <see cref="ByValueMarshalKindSupportDescriptor"/> for pinned by reference parameters. [In] is not allowed. [In, Out] is the default and unnecessary. [Out] is allowed.
+        /// </summary>
+        public static readonly ByValueMarshalKindSupportDescriptor PinnedByReferenceParameterDefault = new ByValueMarshalKindSupportDescriptor(
```

Comment:
This case is still by-value (as we don't support `[In, Out]` for byrefs). Would this name be more accurate?

```suggestion
        public static readonly ByValueMarshalKindSupportDescriptor PinnedArrayParameterDefault = new ByValueMarshalKindSupportDescriptor(
```